### DISCUSSION TITLE
Two different measures for supply

### DISCUSSION THREAD

**Discussion by tsmbland** [EnergySystemsModellingLab/MUSE_OS#321] [General]
I've been working with the `default_timeslices` model, and getting a bit confused as there seems to be two different measures for supply in the results folder which don't match up with each other. You have the `MCAMetric_Supply.csv` file which holds results for all sectors in all years, and also specific files for each year in each sector e.g. `Power/Supply/2025.csv`. The data in these files doesn't match up, for example with the `default_timeslices` model this is what I get for electricity supply in 2025:

![supply_comparison](https://github.com/EnergySystemsModellingLab/MUSE_OS/assets/23723407/aa0a3042-7ee3-4684-ad87-28a7fa230dd1)

Not really sure what's going on here and I need to understand this for [tutorial 7](https://muse-os.readthedocs.io/en/documentation/user-guide/min-max-timeslice-constraints.html). Why are there two different measures for supply? Do they have a different meaning?

**Comment by ahawkes** [comment]
Odd. The MetricMCA_Supply one looks more likely to be correct, as it actually has a profile. Can you check if consumption of electricity matches one or the other ?
A
________________________________
From: Tom Bland ***@***.***>
Sent: Friday, May 31, 2024 2:22:28 PM
To: EnergySystemsModellingLab/MUSE_OS ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [EnergySystemsModellingLab/MUSE_OS] Two different measures for supply (Discussion #321)


This email from ***@***.*** originates from outside Imperial. Do not click on links and attachments unless you recognise the sender. If you trust the sender, add them to your safe senders list<https://spam.ic.ac.uk/SpamConsole/Senders.aspx> to disable email stamping for this address.



I've been working with the default_timeslices model, and getting a bit confused as there seems to be two different measures for supply in the results folder which don't match up with each other. You have the MCAMetric_Supply.csv file which holds results for all sectors in all years, and also specific files for each year in each sector e.g. Power/Supply/2025.csv. The data in these files doesn't match up, for example with the default_timeslices model this is what I get for electricity supply in 2025:

supply_comparison.png (view on web)<https://github.com/EnergySystemsModellingLab/MUSE_OS/assets/23723407/aa0a3042-7ee3-4684-ad87-28a7fa230dd1>

Not really sure what's going on here and I need to understand this for tutorial 7<https://muse-os.readthedocs.io/en/documentation/user-guide/min-max-timeslice-constraints.html>. Why are there two different measures for supply? Do they have a different meaning?

**Comment by tsmbland** [comment]
Consumption of electricity seems to match up with the supply values on the left. Also, there's no discrepancy between consumption values in the `MCAMetric_Consumption.csv` file and the sector/year files:

![consumption_comparison](https://github.com/EnergySystemsModellingLab/MUSE_OS/assets/23723407/4312e621-94db-4a06-86db-82eeff52f06d)


However, the supply values on the right do match up with the very restrictive constraints set in the `TechnodataTimeslices` file for the `default_timeslices` model, where `UtilizationFactor` and `MinimumServiceFactor` are both set to 1 for all timeslices. This essentially means that the technologies must always run at 100% capacity, which would correspond to an annualised supply of 1 and 1.1 in every timeslice for the `gasCCGT` and `windturbine` capacity in 2025. So, actually, the supply values on the right are exactly what you'd expect given these constraints.

It seems like for the model to be successful both of these things need to be true at the same time (i.e. for electricity supply to equal electricity demand, and for the technologies to keep within their supply limits), however in this case clearly both of these things can't happen at the same time given the restrictions we're putting on `MinimumServiceFactor`.

So I'm a bit confused why this model is able to run at all...

**Reply by dalonsoa** [reply]
> It seems like for the model to be successful both of these things need to be true at the same time (i.e. for electricity supply to equal electricity demand, and for the technologies to keep within their supply limits), however in this case clearly both of these things can't happen at the same time given the restrictions we're putting on `MinimumServiceFactor`.

I'm not sure I agree with this. There is no constrain that the electricity supply must equal electricity demand, only that it must _fulfil_ the demand, but can be higher. There's no rule - in this case, at least - forcing the supply to be equal to the demand. So there is no incompatibility here. By setting the production to 100% it fulfills the demand and the minimum service requirement. Most likely, this second constrains is what is wrong, in this case.  

In terms of coding, both things follow different routes to be calculated (why that happens is the question):

1. The files within each sector folder (`Power/Supply/2025.csv` and the like) are calculated with the code in [this function](https://github.com/EnergySystemsModellingLab/MUSE_OS/blob/7c8036caf12509af46e998ea361c34cd0b454e27/src/muse/outputs/sector.py#L198), using the `supply` within the `market` dataset and doing some simple aggregation manipulation within the `market_quantity` function.
2. The `MCA` file comes from [this other function](https://github.com/EnergySystemsModellingLab/MUSE_OS/blob/7c8036caf12509af46e998ea361c34cd0b454e27/src/muse/outputs/mca.py#L349). This calls the [`sector_supply` function](https://github.com/EnergySystemsModellingLab/MUSE_OS/blob/7c8036caf12509af46e998ea361c34cd0b454e27/src/muse/outputs/mca.py#L357), which in turn calls the [`supply` function](https://github.com/EnergySystemsModellingLab/MUSE_OS/blob/7c8036caf12509af46e998ea361c34cd0b454e27/src/muse/quantities.py#L14), each of them more complicated than the previous one. I don't know why the result of this calculation results in a supply identical to the demand - actually, briefly looking at the code, I would expect the supply equal to the maximum production -, but clearly the source of the `supply` in each case is different, so no wonder why the numbers are different.

Now, as I said above, I do not know why there are two sources of information about the supply to be outputted.

**Reply by tsmbland** [reply]
> There is no constrain that the electricity supply must equal electricity demand, only that it must fulfil the demand, but can be higher.

Ok, thanks for clarifying. I guess I had misunderstood what we were constraining in the model. Is this realistic though? (What happens to all the electricity that's produced but not consumed??)

**Reply by dalonsoa** [reply]
That is the question, right? :)

**Comment by ahawkes** [comment]
Yup, sounds like the model should fail to run due to infeasible input data.
A
________________________________

**Comment by tsmbland** [comment]
@dalonsoa @alexdewar

**Comment by ahawkes** [comment]
Right - yes - in reality supply must equal demand for electricity. But often models like this allow supply to be higher as it makes the optimisation faster, and you then rely on convex problem to give you the right answer (i.e. it would be cheaper to supply just the right amount, not more).

But in heavily constrained problems this reliance on convexity often breaks down, so it's a mediocre modelling approach IMO.

In this particular case perhaps the issue is errors in the input data. The user has constrained output to be 100% of capacity. So there is no choice, and supply has to exceed demand?

If yes, can we correct the input data (allow minimum service to be zero) as see?
________________________________

**Comment by dalonsoa** [comment]
I've just found why the discrepancy.

1. The sector output, taking the values right out of the market, gives the right values as market information is calculated based on the constrains (so figure on the right is correct).
2. The MCA information, surprisingly and I don't know why, is obtained by recalculating everything. In [this line](https://github.com/EnergySystemsModellingLab/MUSE_OS/blob/7c8036caf12509af46e998ea361c34cd0b454e27/src/muse/quantities.py#L91), it limits the supply so it is not higher than the maximum production, but it does not impose any minimum constrain but to fulfil the demand, and therefore the minimum service constrained is not honoured.

In summary, the calculation for the `MCAMetric_Supply.csv` is wrong - and I would presume that any other calculation that uses [this supply function](https://github.com/EnergySystemsModellingLab/MUSE_OS/blob/7c8036caf12509af46e998ea361c34cd0b454e27/src/muse/quantities.py#L14) will also be wrong, as no minimum constrains are imposed but to fulfill the demand.

**Comment by dalonsoa** [comment]
See https://github.com/EnergySystemsModellingLab/MUSE_OS/issues/335#issuecomment-2185879445 for an explanation of what's going on.
